File: back-end/src/main/java/es/ull/project/domain/valueobject/ConditionalName.java
package es.ull.project.domain.valueobject;

import java.io.Serializable;

import es.ull.utils.string.UllString;

/**
 * @brief Represents the name of a Conditional. Must be alphanumeric, non-null, and under max size.
 */
public final class ConditionalName implements Serializable {

    private static final String ERROR_NAME_NULL = "Name for conditional is not defined";
    private static final String ERROR_NAME_EMPTY = "Name for conditional is empty";
    private static final String ERROR_NAME_FORMAT = "Invalid name for conditional. Only alphanumeric characters are allowed.";
    private static final String ERROR_NAME_SIZE = "Size for conditional name is too long";

    private static final int MAX_NAME_SIZE = 50;

    private final String name;

    public ConditionalName(String name) {
        validateName(name);
        this.name = name;
    }

    private void validateName(String name) {
        if (name == null) {
            throw new IllegalArgumentException(ERROR_NAME_NULL);
        }
        name = name.trim();
        if (name.isEmpty()) {
            throw new IllegalArgumentException(ERROR_NAME_EMPTY);
        }
        if (name.length() > MAX_NAME_SIZE) {
            throw new IllegalArgumentException(ERROR_NAME_SIZE);
        }
        if (!UllString.isAlphanumeric(name)) {
            throw new IllegalArgumentException(ERROR_NAME_FORMAT);
        }
    }

    public String getName() {
        return this.name;
    }

    public ConditionalName setName(String newName) {
        return new ConditionalName(newName);
    }

    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) return true;
        if (!(otherObject instanceof ConditionalName)) return false;
        final ConditionalName that = (ConditionalName) otherObject;
        return this.name.equals(that.name);
    }

    @Override
    public int hashCode() {
        return this.name.hashCode();
    }

    @Override
    public String toString() {
        return String.format("ConditionalName={%s}", this.name);
    }
}

File: front-end/src/domain/value-object/conditional-name.ts
/**
 * Represents the name of a Conditional.
 */
export class ConditionalName {
  private static readonly ERROR_NAME_NULL = 'Name for conditional is not defined';
  private static readonly ERROR_NAME_EMPTY = 'Name for conditional is empty';
  private static readonly ERROR_NAME_FORMAT = 'Invalid name for conditional. Only alphanumeric characters are allowed.';
  private static readonly ERROR_NAME_TOO_LONG = 'Size for conditional name is too long';

  private static readonly MAX_LENGTH = 50;

  private readonly name: string;

  constructor(name: string) {
    this.validateName(name);
    this.name = name;
  }

  private validateName(name: string): void {
    if (name === null || name === undefined) {
      throw new Error(ConditionalName.ERROR_NAME_NULL);
    }

    const trimmed = name.trim();

    if (trimmed.length === 0) {
      throw new Error(ConditionalName.ERROR_NAME_EMPTY);
    }

    if (trimmed.length > ConditionalName.MAX_LENGTH) {
      throw new Error(ConditionalName.ERROR_NAME_TOO_LONG);
    }

    if (!/^[a-zA-Z0-9]+$/.test(trimmed)) {
      throw new Error(ConditionalName.ERROR_NAME_FORMAT);
    }
  }

  public getName(): string {
    return this.name;
  }

  public setName(newName: string): ConditionalName {
    return new ConditionalName(newName);
  }

  public toString(): string {
    return `ConditionalName={${this.name}}`;
  }

  public equals(other: ConditionalName): boolean {
    return this.name === other.getName();
  }
}
